### PR TITLE
Single setup py

### PR DESCRIPTION
**Pull request by toobaz**
OK, this is related to #114 .

I was a bit drastic... I won't get offended if you want to reject something ;-)

Anyway:
```
python setup.py build_ext --inplace
python3 setup.py build_ext --inplace
```
will now generate the automatically generated .pyx _and_ the .c files _if Cython is installed_. If it's not, they will fallback to the existing .c files (which are assumed to be shipped in the release tarball - according with ``MANIFEST.in``). I ran the tests and everything seems to work fine.

By the way,
```
python setup.py build_sphinx
```

will generate the documentation. This does not work with python3 because of a SyntaxError I didn't bother to fix (fixing it is trivial, keeping compatibility with both python 2 and 3 is not, at least for me).

### CODE REVIEW

**Comment by kwgoodman** [comment]
The PR is failing on travis (probably because our travis.yml file does not install cython and the auto-gen c files are not in the repo anymore [which is great!]): https://travis-ci.org/kwgoodman/bottleneck/builds/132676230

The top-level makefile got deleted. I need that one. The doc makefile too. Good catch on the py3 compat of the docs. I guess we should add a unit test that makes the docs (we can keep that separate from this PR---you've got enough on your hands).

**Comment by kwgoodman** [comment]
Oh, sorry, nevermind about the makefiles. Very nice that you got rid of them. I'll add back a smaller top-level makefile later.

**Comment by toobaz** [comment]
Concerning the makefiles: I told you I was a bit drastic! ;-) Feel free to filter my changes.

Concerning Travis, yes, your diagnosis is correct. I never played with Travis however. I can try to add a ``conda install cython`` at the end of ``ci/travis/conda_setup.sh``...

**Comment by kwgoodman** [comment]
@Midnighter to add cython to travis do we just add cython to
```
env: DEPS="numpy==1.11.0 nose" PYTHON_VERSION="2.7" PYTHON_ARCH="64"
```
in .tavis.yml?

**Comment by Midnighter** [comment]
Exactly, just put all required packages in the `DEPS="..."` string and conda will install them when the environment is generated.

Btw, I'm a big fan of the following [cookiecutter](https://github.com/ionelmc/cookiecutter-pylibrary) that generates a library project skeleton for you. There the docs are also only generated with Python 2.7. Seems to be a general thing.

**Comment by toobaz** [comment]
Uhm... now I'm clueless.

**Comment by Midnighter** [comment]
For extensions the option is called `include_dirs=np.get_include()` is that it instead of `include_path`?

**Comment by kwgoodman** [comment]
It is now passing on linux but not osx. On osx the moving window functions don't get built. Unfortunately travis [does not show](https://travis-ci.org/kwgoodman/bottleneck/jobs/132884953#L248) the compiling of bottleneck. Is there anyway to get travis to show it? Should look something like:

```
kg@kg:/devel/bottleneck (unroll *)$ ma
rm -rf build dist Bottleneck.egg-info
find . -name \*.pyc -delete
rm -rf bottleneck/src/auto_pyx/*.c bottleneck/src/auto_pyx/*.html bottleneck/src/auto_pyx/build bottleneck/src/auto_pyx/../../*.so
python -c "from bottleneck.src.template.template import make_pyx; make_pyx();"
cython bottleneck/src/auto_pyx/reduce.pyx
cython bottleneck/src/auto_pyx/nonreduce.pyx
cython bottleneck/src/auto_pyx/nonreduce_axis.pyx
cython bottleneck/src/auto_pyx/move.pyx
rm -rf bottleneck/src/auto_pyx/../reduce.so
python bottleneck/src/auto_pyx/reduce_setup.py build_ext --inplace
running build_ext
skipping 'bottleneck/src/auto_pyx/reduce.c' Cython extension (up-to-date)
building 'reduce' extension
creating build
creating build/temp.linux-x86_64-2.7
creating build/temp.linux-x86_64-2.7/bottleneck
creating build/temp.linux-x86_64-2.7/bottleneck/src
creating build/temp.linux-x86_64-2.7/bottleneck/src/auto_pyx
x86_64-linux-gnu-gcc -pthread -fno-strict-aliasing -DNDEBUG -g -fwrapv -O2 -Wall -Wstrict-prototypes -fPIC -I/usr/local/lib/python2.7/dist-packages/numpy/core/include -I/usr/include/python2.7 -c bottleneck/src/auto_pyx/reduce.c -o build/temp.linux-x86_64-2.7/bottleneck/src/auto_pyx/reduce.o
In file included from /usr/local/lib/python2.7/dist-packages/numpy/core/include/numpy/ndarraytypes.h:1777:0,
                 from /usr/local/lib/python2.7/dist-packages/numpy/core/include/numpy/ndarrayobject.h:18,
                 from /usr/local/lib/python2.7/dist-packages/numpy/core/include/numpy/arrayobject.h:4,
                 from bottleneck/src/auto_pyx/reduce.c:266:
<snip>
```

That should show us the error which would be nice since I don't have osx.

**Comment by toobaz** [comment]
Although everything worked here, I kept in me a sense of guilt for having removed that flag! Now I feel better.

**Comment by kwgoodman** [comment]
Looks good. Could you make some tweaks?

- put `bottleneck/template/__init__.py` back into `bottleneck/template/template.py`
- revive the top-level makefile, updating it to use the setup.py file (which will cut out a lot of junk)

**Comment by kwgoodman** [comment]
tox no longer runs:
```
kg@kg:/devel/bottleneck (toobaz-single_setup_py)$ tox
<snip>
bottleneck/move.c:287:30: fatal error: csrc/move_median.c: No such file or directory
 #include "csrc/move_median.c"
<snip>
```

It is because the csrc directory doesn't appear in the sdist.

**Comment by Midnighter** [comment]
> On osx the moving window functions don't get built. Unfortunately travis does not show the compiling of bottleneck. Is there anyway to get travis to show it?

At the moment the `.travis.yml` contains the line `pip install .` which swallows that output. If you change that to `python setup.py install` instead (or `build_ext`) the output should be there.

File: .travis.yml [hunk 1]
```diff
@@ -3,17 +3,17 @@ sudo: false
 matrix:
     include:
         - os: linux
-          env: DEPS="numpy==1.11.0 nose" PYTHON_VERSION="2.7" PYTHON_ARCH="64"
+          env: DEPS="numpy==1.11.0 nose cython" PYTHON_VERSION="2.7" PYTHON_ARCH="64"
```
Comment (Midnighter) [inline .travis.yml]: Since you're pinning the numpy version it might be a good idea to do that for Cython as well.

**Comment by kwgoodman** [comment]
On travis we are currently only testing installation on a system that has cython installed. Would be nice to also test without cython. To do so I guess we'd have to make an sdist and then install that. Not sure how to do that. Point pip at the sdist archive might work.

That would double our number of environments to 12. Maybe we could drop python 3.4? Which would give us 8.

We can do this is a separate PR.

**Comment by Midnighter** [comment]
For the separate pull request:

1. I didn't find it earlier but `pip` has a `--verbose` option which gives you all the output that you wanted to see.
2. Running `python setup.py sdist` creates an archive like this: `dist/Bottleneck-1.1.0.dev0.tar.gz` which can then be installed with `pip`. I'd be happy to make a pull request that adapts those changes in the `.travis.yml` once you're satisfied with the state of the master branch.

**Pull request merged**